src/main/java/pl/psnc/dei/service/BatchService.java

```diff
@@ -306,4 +307,78 @@ private String extractIdentifier(String key) {
 		Hibernate.initialize(currentProject.getDatasets());
 		return currentProject.getDatasets();
 	}
-}
+
+	/**
+	 * Saves records to DB, and calculate difference of saved and provided ones, as some of them can
+	 * be omitted during persiting process
+	 * @param projectName name of project to which data should be saved
+	 * @param datasetName name of dataset from which data comes
+	 * @param recordsIds id of data to save
+	 * @return saved records
+	 * @throws NotFoundException
+	 */
+	public Set<Record> uploadRecordsToProject(String projectName,
+											  String datasetName,
+											  Set<String> recordsIds)
+			throws NotFoundException {
+		Optional<Dataset> optionalDataset = this.datasetsReposotory.findByName(datasetName);
+		if(optionalDataset.isEmpty()) {
+			throw new NotFoundException("Dataset with name: " + datasetName + " not found");
+		}
+		String datasetId = optionalDataset.get().getDatasetId();
+		Set<Record> records = this.uploadRecords(projectName, datasetId, recordsIds);
+		if (records.size() < recordsIds.size()) {
+			String difference = records
+					.stream()
+					.filter(record -> !recordsIds.contains(record.getIdentifier()))
+					.map(Record::getIdentifier).collect(Collectors.joining(","));
+			if (difference.isEmpty()) {
+				difference = recordsIds.stream().collect(Collectors.joining(","));
+			}
+			log.warn("Following records will not be added to the import: {}", difference);
+		}
+		return records;
+	}
+
+	public List<Import> makeComplexImport(MultipartFile file, String name, String projectName, String datasetName) throws IOException {
+		File saved = File.createTempFile("tmp", ".csv");
+		file.transferTo(saved);
+		List<Import> result = this.makeComplexImport(saved, name, projectName, datasetName);
+		saved.delete();
+		return result;
+	}
+
+	public List<Import> makeComplexImport(InputStream inputStream, String name, String projectName, String datasetName) throws IOException {
+		File saved = File.createTempFile("tmp", ".csv");
+		OutputStream os = new FileOutputStream(saved);
+		os.write(inputStream.readAllBytes());
+		List<Import> result = this.makeComplexImport(saved, name, projectName, datasetName);
+		saved.delete();
```

Comment:
same situation with cleaning up temp stuff in case of execution termination.

Reply:
same as above